modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/dtx/DtxBranch.java

```diff
@@ -0,0 +1,225 @@
+        if (null != state && state == State.ACTIVE) {
+            associatedSessions.put(sessionId, State.SUSPENDED);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    public void enqueueMessage(AndesMessage andesMessage) {
+        enqueueList.add(andesMessage);
+    }
+
+    public void enqueueMessages(Collection<AndesMessage> messagesList) {
+        enqueueList.addAll(messagesList);
+    }
+
+    public boolean hasAssociatedActiveSessions() {
+        if (hasAssociatedSessions()) {
+            for (State state : associatedSessions.values()) {
+                if (state != State.SUSPENDED) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    private boolean hasAssociatedSessions() {
+        return !associatedSessions.isEmpty();
+    }
+
+    public void clearAssociations() {
+        associatedSessions.clear();
+    }
+
+    public boolean expired() {
+        // TODO implement transaction timeouts
+        return false;
+    }
+
+    public State getState() {
+        return state;
+    }
+
+    public void prepare() throws AndesException {
+        LOGGER.debug("Performing prepare for DtxBranch {}" + xid);
+        internalXid = dtxRegistry.storeRecords(xid, enqueueList, dequeueList);
+    }
+
+    public void setState(State state) {
+        this.state = state;
+    }
+
+    public void dequeueMessages(List<AndesAckData> ackList) {
+        dequeueList.addAll(ackList);
+    }
+
+    public void rollback() throws AndesException {
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("Performing rollback for DtxBranch {}" + xid);
+        }
+
+        if (internalXid != NULL_XID) {
+            dtxRegistry.removePreparedRecords(internalXid);
+            internalXid = NULL_XID;
+        } else {
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug("Cannot rollback since could not find a internal XID " + "{}" + xid);
+            }
+        }
+    }
+
+    public void commit(Runnable callback, AndesChannel channel) throws AndesException {
+        this.callback = callback;
+        eventManager.requestDtxCommitEvent(this, channel);
+    }
+
+    @Override
+    public void updateState() throws AndesException {
+        SlotMessageCounter.getInstance().recordMetadataCountInSlot(enqueueList);
+        enqueueList.clear();
+        dequeueList.clear();
+        callback.run();
+
+        LOGGER.debug("Dtx commit messages state updated. Internal Xid " + internalXid);
+        // TODO: Handle exceptions
+    }
+
+    @Override
+    public String eventInfo() {
+        return null;
+    }
+
+    public void clearEnqueueList() {
+        enqueueList.clear();
+    }
+
+    /**
+     * Getter for internalXid
+     */
+    public long getInternalXid() {
+        return internalXid;
+    }
+
+    public enum State {
+        SUSPENDED, ACTIVE, ROLLBACK_ONLY, PREPARED, FORGOTTEN, TIMED_OUT, HEUR_COM, HEUR_RB
```

Comment:
Good if we could add comments mentioning the role of each state and it's applicability